_posts/2024-06-24-solarkraft-part4.md

```diff
@@ -0,0 +1,200 @@
+
+Taking a step back from the concrete solution, let's try to answer the more abstract question: _What do we want to achieve with runtime monitors in blockchains?_ As eventually runtime monitors are going to be deployed and executed _on the live blockchain_, and they should satisfy the three requirements:
+
+- **Prevent safety violations** (_safety_): bad things, such as your token balance being stolen, should not happen. This is the primary goal of runtime monitors: react preventively, and abort unwanted executions.
+- **Detect liveness violations** (_liveness_): good things should be able to happen! E.g. you, as an account owner, should be able to withdraw your own balance. If a legitimate transaction keeps reverting, that's also a bug, not less severe than someone stealing your tokens.
+- **Detect unexpected behaviors** (_completeness_): same as code, specs are not perfect. If a spec author overlooked some behaviors, and they manifest themselves on the live blockchain, this may mean anything: from simple spec incompleteness, to an exploit being executed. Monitors should be configurable to either issue a simple warning, or to prevent such behaviors altogether.
+
+The problem we've observed with the previously employed specification approaches is that the specs of _what_ the method should do can easily be much larger than the actual implementation; compare e.g. this [ERC-721 Certora spec][] with the [ERC-721 implementation][] (don't forget to exclude comments when comparing). So we would like to add to the above the following soft requirement:
+
+- **Specify behaviors compactly and independently** (_compactness and modularity_): it is usually the case that a smart contract encompasses a lot of various aspects (e.g. authentication, authorization, storage management, math computations), as well is written/employed/reasoned about by various roles (e.g. smart contract developer, mathematician, architect, UI developer). All of those roles should be able to specify various aspects of the smart contract behavior as easily and as independently as possible.
+
+So monitors should be able to specify both _safety_ and _liveness_ properties, be _complete_ wrt. the current and future system behaviors, and, preferably, also be _compact and modular_. For that we propose to apply a certain structure to the _direct monitors_ (those reasoning from cause to effect), as well as complement it with _reverse monitors_ (those going from effect to cause), thus combining them together in what we call _hybrid monitors_.
+
+
+## Direct Monitors
+
+Here we reason from the cause (method invocation) to the effect, but apply a structure which closely mimics in formal semantics what we expect to see when we program smart contracts. The essence of the structure is in the picture below:
+
+![Direct monitor specs](/img/DirectMonitors.png)
+
```

Comment:
```suggestion
In other words, we distinguish three kinds of properties:
```

Reply:
makes sense; but I will make it "three kinds of conditions"; "properties" are too general for that...